### DISCUSSION TITLE
Is netfox suited for a FPS high-paced game despite using gdscript?

### DISCUSSION THREAD

**Discussion by jokoon** [foxssake/netfox#231] [General]
I heard about this library, which is a pain to use since it requires to rebuild godot, and it is not available as a gdextension (yet?):

https://github.com/GameNetworking/NetworkSynchronizer

I want network reconciliation for my FPS game, and honestly I don't know enough implementation details and internals to know if it's really worth it to this one just because it's made in C++.

I read somebody on discord say this:

> There's a few add-on frameworks that directly implement it using the native multiplayer stack and node pattern. 
> 
> They're in Gdscript but if you look at what they're doing its not anything highly computational. They're basically instructing the C++ engine in Godot to resimulate and update everyone in positions.

Would you mind explaining if the C++ module above is faster, and if there are limitations to netfox?

Is netfox "fast enough" despite using gdscript?

**Comment by elementbound** [comment]
Hi @jokoon ,

The quote sums up netfox correctly imo - most of what netfox does is just move data around, so it's not that intense on CPU. But it really depends on your project.

Haven't tested it yet, but imo you should be fine with up to 8 players easily. More should be fine too, but I'd definitely start measuring if you expect higher player counts. My personal experience is that I could sync 2 players with RollbackSynchronizer and ~150 NPCs using StateSynchronizer, and netfox still wouldn't be a performance bottleneck ( simulating 150 NPCs was though :P )

For FPS games specifically, what you could have issues with is raycasting - you can rewind the game state to the point of fire, but not sure if the physics system would pick up the rewind before you do the raycast. Make sure to check before comitting to a bigger project. Note that this is a limitation of Godot, not netfox.

**Reply by singerbj** [reply]
@elementbound any creative ideas on how one could work around this godot limitation? It seems the pr is stuck in purgatory never to be merged https://github.com/godotengine/godot/pull/76462

**Reply by elementbound** [reply]
Good news @singerbj! There's support for physics now: https://foxssake.github.io/netfox/latest/netfox.extras/guides/physics/
Your two options are either using a custom build of the PR you mentioned ( there's a community-run build linked in the docs ), or use the Rapier addon, which enables manually updating and stepping the physics sim.

If it's strictly only for raycasting, there's also a workaround here: https://foxssake.github.io/netfox/latest/netfox/tutorials/rollback-caveats/#physics-updates

At the moment I don't have much info on which one is more performant. Personally I'd check if I could get away with the workaround, as that's less intrusive than rollback physics.

**Reply by singerbj** [reply]
@elementbound thanks so much for this info! I will check out the workaround!

**Comment by jokoon** [comment]
I also plan to implement bullet ballistics.

Originally I wanted to implement with CCD continuous collision detection,
but I decided to use frame by frame raycast instead.

I assume that godot would move those bullet as godot objects, in a
deterministic fashion.
